rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,962 @@
+    demangle_type,
+    true,
+    allocator,
+    endpoints_info);
+}
+
+using NamesAndTypes = std::map<std::string, std::set<std::string>>;
+
+static
+void
+__get_names_and_types(
+  const GraphCache::EntityGidToInfo & entities,
+  DemangleFunctionT demangle_topic,
+  DemangleFunctionT demangle_type,
+  NamesAndTypes & topics)
+{
+  assert(nullptr != demangle_topic);
+  assert(nullptr != demangle_type);
+  for (const auto & item : entities) {
+    std::string demangled_topic_name = demangle_topic(item.second.topic_name);
+    if ("" != demangled_topic_name) {
+      topics[demangled_topic_name].insert(demangle_type(item.second.topic_type));
+    }
+  }
+}
+
+static
+rmw_ret_t
+__populate_rmw_names_and_types(
+  NamesAndTypes topics,
+  rcutils_allocator_t * allocator,
+  rmw_names_and_types_t * topic_names_and_types)
+{
+  if (topics.empty()) {
+    return RMW_RET_OK;
+  }
+
+  rmw_ret_t rmw_ret = rmw_names_and_types_init(topic_names_and_types, topics.size(), allocator);
+  if (RMW_RET_OK != rmw_ret) {
+    return rmw_ret;
+  }
+
+  size_t index = 0;
+  for (const auto & item : topics) {
+    char * topic_name = rcutils_strdup(item.first.c_str(), *allocator);
+    if (!topic_name) {
+      RMW_SET_ERROR_MSG("failed to allocate memory for topic name");
+      rmw_ret = RMW_RET_BAD_ALLOC;
+      goto cleanup;
+    }
+    topic_names_and_types->names.data[index] = topic_name;
+
+    {
+      rcutils_ret_t rcutils_ret = rcutils_string_array_init(
+        &topic_names_and_types->types[index],
+        item.second.size(),
+        allocator);
+      if (RCUTILS_RET_OK != rcutils_ret) {
+        RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+        rmw_ret = rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+        goto cleanup;
+      }
+    }
+    size_t type_index = 0;
+    for (const auto & type : item.second) {
+      char * type_name = rcutils_strdup(type.c_str(), *allocator);
+      if (!type_name) {
+        RMW_SET_ERROR_MSG("failed to allocate memory for type name");
+        rmw_ret = RMW_RET_BAD_ALLOC;
+        goto cleanup;
+      }
+      topic_names_and_types->types[index].data[type_index] = type_name;
+      ++type_index;
+    }
+    ++index;
+  }
+  return RMW_RET_OK;
+cleanup:
+  if (RMW_RET_OK != rmw_names_and_types_fini(topic_names_and_types)) {
+    RCUTILS_LOG_ERROR_NAMED(
+      log_tag,
+      "error during report of error: %s", rmw_get_error_string().str);
+  }
+  return rmw_ret;
+}
+
+rmw_ret_t
+GraphCache::get_names_and_types(
+  DemangleFunctionT demangle_topic,
+  DemangleFunctionT demangle_type,
+  rcutils_allocator_t * allocator,
+  rmw_names_and_types_t * topic_names_and_types) const
+{
+  assert(demangle_topic);
+  assert(demangle_type);
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(allocator, "get_node_names allocator is not valid",
+    return RMW_RET_INVALID_ARGUMENT);
+  if (RMW_RET_OK != rmw_names_and_types_check_zero(topic_names_and_types)) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  NamesAndTypes topics;
```

Comment:
@ivanpauno nit: add a comment explaining why the intermediate representation (i.e. easier using an RB tree or hash map).